### ISSUE TITLE
GDTTransformer Can't endBackgroundTask. 

### ISSUE THREAD

**Issue by zhangcx627** [firebase/firebase-ios-sdk#6258]
* Xcode version: Xcode 11.4 (11E146)
  * Installation method: CocoaPods

#### Steps to reproduce:
Don't know which component crash. We used pod to get firebase like this.
  
> pod 'FirebaseInstanceID'
>   pod 'Firebase/Analytics'
>   pod 'Firebase/RemoteConfig'
>   pod 'Firebase/Crashlytics'
>   pod 'Firebase/Performance'
>   pod 'Firebase/InAppMessaging'

When my phone upgrade to iOS 14 Beta,  and app enter background. Later it will crash.  Crash log is 
`[BackgroundTask] Background task still not ended after expiration handlers were called: <_UIBackgroundTaskInfo: 0x2835e06c0>: taskID = 64, taskName = CCTLogWriter, creationTime = 64357 (elapsed = 0). This app will likely be terminated by the system. Call UIApplication.endBackgroundTask(_:) to avoid this.
Message from debugger: Terminated due to signal 9`
And when I use `UIApplicationEndBackgroundTaskError()` to debug where goes wrong,  It breakpoint here.
![F7D914F9D38C4B9558FE20619DC154FE](https://user-images.githubusercontent.com/20290940/89985459-b279c880-dcad-11ea-98a9-44da326565ea.jpg)

Could you please check this method to see if the backgroundtask is closed?

**Comment by maksymmalyhin** [comment]
@zhangcx627 Thank you for the report. We will investigate the issue.

**Comment by maksymmalyhin** [comment]
@zhangcx627 Could you please share `Podfile.lock` so we can know for sure what component versions are used?

**Comment by maksymmalyhin** [comment]
It looks like an old version of GoogleDataTransport is used in your app, because a background task with name `CCTLogWriter` is not started in the recent versions. Could you please try to update Firebase and check if you still have an issue.

**Comment by morganchen12** [comment]
Likely related to or a duplicate of #4554

**Comment by zhangcx627** [comment]
@maksymmalyhin  Thank you~. I'm try to update to the last version and use Xcode 12 Beta4. And find the log about `[BackgroundTask] Background task still not ended after expiration handlers were called: <_UIBackgroundTaskInfo: 0x2835e06c0>: taskID = 64, taskName = CCTLogWriter, creationTime = 64357 (elapsed = 0). This app will likely be terminated by the system. Call UIApplication.endBackgroundTask(_:) to avoid this.` disappear. But it still crash `Message from debugger: Terminated due to signal 9`. I don’t know if there are any more with firebase, no matter what, thank you very much~

**Comment by maksymmalyhin** [comment]
@zhangcx627 I suspect, there may be something that prevents GoogleDataTransport (the SDK that contained `CCTLogWriter ` in older versions) from updating. If you could share your `Podfile.lock` we can take a look together what it may be.

**Comment by zhangcx627** [comment]
Sorry about the slow reply. Here is the Podfile.lock.
[Podfile.lock.zip](https://github.com/firebase/firebase-ios-sdk/files/5082434/Podfile.lock.zip)

**Comment by zhangcx627** [comment]
May be it's Realm's bug. 
https://github.com/realm/realm-cocoa/issues/6671

**Comment by zhangcx627** [comment]
@maksymmalyhin Does this log from the previous versions?
`[BackgroundTask] Background Task 55 ("GDTTransformer"), was created over 30 seconds ago. In applications running in the background, this creates a risk of termination. Remember to call UIApplication.endBackgroundTask(_:) for your task in a timely manner to avoid this.`

**Comment by maksymmalyhin** [comment]
@zhangcx627 Sorry for the slow response, I was off last week. Thank you for the update.

Yeah `GDTTransformer` background task ID is used in the new versions. Do you still experience the crash with the recent Firebase version? If yes, will it be possible to share the crash log or at least stack traces for all threads where Firebase or GDT is present? 

Also, a sample project to reproduce the issue will help a lot. Would you be able to post a sample project?

**Comment by zhangcx627** [comment]
No, there is no crash, But there are just a lot of logs `[BackgroundTask] Background Task 55 ("GDTTransformer"), was created over 30 seconds ago. In applications running in the background, this creates a risk of termination. Remember to call UIApplication.endBackgroundTask(_:) for your task in a timely manner to avoid this.`

**Comment by maksymmalyhin** [comment]
@zhangcx627 Thank you for the update. I was not able to reproduce the issue on my end. I haven't detected anything obviously wrong in the implementation either. 
I guess, there are some additional conditions to trigger the issue. Could you please provide a sample project and steps to reproduce the issue so we can make progress on it?

**Comment by zhangcx627** [comment]
Verrrrrrrrry sorry about late reply. This is the demo. Please change the keys in GoogleService-Info.plist first.
Use Xcode 12.0 beta6 build and use iOS14 Device not simulator. 
If this log doesn't appear, please switch front and back sometimes. It will appear. 
My app will have this log when I switch to the background for the first time

[FBDemo.zip](https://github.com/firebase/firebase-ios-sdk/files/5181059/FBDemo.zip)

**Comment by maksymmalyhin** [comment]
@zhangcx627 Thank you for the provided project. I could get the ` ("GDTTransformer"), was created over 30 seconds ago...`  console message only after generating a big amount (more than 10) Performance traces in the background. In this case the background tasks were successfully finished in the expiration handler which is an expected behaviour and should not lead to an app crash.

I'll take a look if any optimization can be provided to increase chance of finishing the work before the background task expiration.

From your side I would suggest to check your app for extra Performance traces, especially ones that may be created near the transition to the background time.

**Comment by jason-oanda** [comment]
Hey folks,

I'm also getting this same error:
`2020-09-08 17:02:24.528971-0400 iosapp[23130:550346] [BackgroundTask] Background Task 4615 ("GDTTransformer"), was created over 30 seconds ago. In applications running in the background, this creates a risk of termination. Remember to call UIApplication.endBackgroundTask(_:) for your task in a timely manner to avoid this.`

In my case it's not related to iOS 14 (I'm building/running on iOS 13.7 from Xcode 11.7) but rather started happening after adding the Firebase Performance framework via CocoaPods.

I can reproduce both on the Simulator and on device (iPhone 11 Pro). After building/running, launching the app and then waiting about 3-4 minutes the log messages start coming in without fail.

I reverted the change (removed Firebase/Performance from Podfile), did 'pod install' (which removed FirebasePerformance, GTMSessionFetcher, GoogleToolboxForMac) rebuilt the project, and now I can't reproduce.

[Podfile.lock with Performance](https://github.com/firebase/firebase-ios-sdk/files/5191222/Podfile.lock-with-Performance.zip) / [Podfile.lock without Performance](https://github.com/firebase/firebase-ios-sdk/files/5191223/Podfile.lock-without-Performance.zip)

This probably isn't very helpful but seems related. Please let me know (and forgive me) if you think this is sufficiently different from the issue description such that it should be reported as its own issue.

Thank you!

**Comment by zhangcx627** [comment]
Thank you all~~~. Sorry about the issue title. The first time I noticed this problem was when adapting to iOS14. And later found that crash is from Realm. Now there only console those logs. My app has lots of background traces. So it must appear when I debug.  
`I reverted the change (removed Firebase/Performance from Podfile), did 'pod install' (which removed FirebasePerformance, GTMSessionFetcher, GoogleToolboxForMac) rebuilt the project, and now I don't see the GDTTransformer logs. Adding Firebase/Performance causes them to reappear.`
I will try this next. 
Thank you again.

**Comment by maksymmalyhin** [comment]
Looks like background tasks are not properly finished by `GDTTransformer` for events filtered by a transformer [here](https://cs.opensource.google/firebase-sdk/firebase-ios-sdk/+/master:GoogleDataTransport/GDTCORLibrary/GDTCORTransformer.m;l=73;drc=85e33543469ec7e55b0da8153d1d38761d3848aa) (in this case Firebase Performance event rate limit filter applied). I'll work on a fix.

**Comment by maksymmalyhin** [comment]
The fix is planned for the next release.

**Comment by zhangcx627** [comment]
Thanks very much.

**Comment by icedice** [comment]
Has this been released? I'm on 6.33.0 (from Carthage) and keep seeing logs like this:

`2020-09-30 16:08:15.383822+0200 XXXX[11617:3146215] [BackgroundTask] Background Task 13 ("GDTTransformer"), was created over 30 seconds ago. In applications running in the background, this creates a risk of termination. Remember to call UIApplication.endBackgroundTask(_:) for your task in a timely manner to avoid this.`

Releasing with this does not seem safe...

**Comment by maksymmalyhin** [comment]
@icedice This is supposed to be release. Let me try to recreate it on my end. In the meanwhile I would recommend to double check that Carthage frameworks were actually updated with the proper version and successfully copied to you project.

**Comment by maksymmalyhin** [comment]
@icedice  I cannot reproduce it locally in a new project, so I would suggest to clean Carthage cache, double check that the frameworks were updated properly and clean project to make sure it is not some weird caching issue before doing other troubleshooting steps.

**Comment by icedice** [comment]
I tried deleting the whole Carthage folder along with the whole cartage cache folder and then bootstrapping again. 
I'm still seeing the issue in both simulator and on actual device. 

here is my cartfile.resolver (relevant part):

```
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseABTestingBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAuthBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseCrashlyticsBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseFirestoreBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseProtobufBinary.json" "6.33.0"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseRemoteConfigBinary.json" "6.33.0"

```
I tried inspecting the .framework for a version number but the result looks strange:

```
plutil -p /Carthage/Build/iOS/FirebaseFirestore.framework/info.plist                                                                                                                                                                                                                                                                                                  
{
  "CFBundleExecutable" => "FirebaseFirestore"
  "CFBundleIdentifier" => "com.firebase.Firebase-FirebaseFirestore"
  "CFBundleInfoDictionaryVersion" => "6.0"
  "CFBundleName" => "FirebaseFirestore"
  "CFBundlePackageType" => "FMWK"
  "CFBundleVersion" => "1"
  "DTSDKName" => "iphonesimulator11.2"
}
```

Could this in anyway be related to us applying this fix: https://github.com/Carthage/Carthage/issues/3019#issuecomment-699143260 to accommodate for the Carthage / XC12 issue? As Carthage is simply downloading the binaries for firestore my guess is that is should not be related...
